contracts/utils/structs/Heap.sol

```diff
@@ -0,0 +1,570 @@
+// SPDX-License-Identifier: MIT
+// This file was procedurally generated from scripts/generate/templates/Heap.js.
+
+pragma solidity ^0.8.20;
+
+import {SafeCast} from "../math/SafeCast.sol";
+import {Comparators} from "../Comparators.sol";
+import {Panic} from "../Panic.sol";
+
+/**
+ * @dev Library for managing https://en.wikipedia.org/wiki/Binary_heap[binary heap] that can be used as
+ * https://en.wikipedia.org/wiki/Priority_queue[priority queue].
+ *
+ * Heaps are represented as an array of Node objects. This array stores two overlapping structures:
+ * * A tree structure where the first element (index 0) is the root, and where the node at index i is the child of the
```

Comment:
if we use index 1 instead of 0 it can simplify operations of parent, left-child and right-child from (n-1)/2, 2n+1 and 2n+2 to n/2, 2n and 2n + 1, respectively.

this makes parent and left-child more optimal. if p points to a 0-based array and q = p - 1 then we can access p[0] as q[1] so there is no overhead in using a 1-based array

Reply:
This makes is so that `array.length` is no longer the size of the Heap. It also creates a possible issue with empty array and array with one elements both being somehow empty.

The complexity of doing `(n-1)/2` or `2n+1`, `2n+2` is not that much more complexe compared to the special cases we would have to consider